src/terrain/services/permanent_id_requests.clj

```diff
@@ -184,46 +185,102 @@ If this dataset accompanies a paper, please contact us with the DOI for that pap
 (defn- move-folder
   [user {:keys [id path]} dest-path]
   (try+
-   (data-info-client/move-single user id dest-path)
-   (ft/path-join dest-path (ft/basename path))
-
+   (-> (data-info-client/move-single user id dest-path)
+       :body
+       (json/decode true)
+       :async-task-id)
    (catch Object e
      (log/error e)
      (email/send-permanent-id-request-data-move-error path dest-path current-user (:body e (str e)))
-     path)))
+     nil)))
+
+
+(defn- async-move-poller
+  [{:keys [id path]} dest-path async-task-id post-move-fn]
+  (try+
+    (log/info "Waiting for async move to complete for" id path)
+
+    (loop [sleep-seconds 1]
+      (Thread/sleep (* sleep-seconds 1000))
+
+      (let [{:keys [statuses end_date]} (async-tasks-client/get-by-id async-task-id)
+            last-status (-> statuses last :status)]
+        (when (and end_date (not= last-status "completed"))
+          (log/warn "Async folder move not completed successfully:" last-status))
+
+        (if end_date
+          ;; Async move completed.
+          (post-move-fn (:path (data-info/stat-by-uuid (config/irods-user)
+                                                       id
+                                                       :filter-include "path")))
+          ;; Else keep waiting for up to wait-seconds-max.
+          ;; Since the sleep seconds are doubled each time,
+          ;; then this thread has already been waiting
+          ;; ((sleep-seconds * 2) - 1) number of seconds.
+          (if (< (dec (* sleep-seconds 2)) (config/permanent-id-async-move-wait-seconds-max))
+            (recur (* sleep-seconds 2))
+            (throw+ (str "Async move took too long to complete successfully. Last status: "
+                         last-status))))))
```

Comment:
Hopefully these comments make sense.

Reply:
I had to do a bit of mental math to verify the claim that at this point in the loop it'll have slept `(- (* sleep-seconds 2) 1)` seconds, but once I'd satisfied my own curiosity it makes sense. I'm imagining it's not super important that we fit in exactly under the maximum, since this can end up running past it (but only one loop, I think), but mostly just that we don't let it loop forever.

Reply:
Yeah, I had to think through the math a couple of times before I got it right. In case it's not obvious to everyone, this come from one of those sum formulas I remember from my CS courses: `∑2^n from 0 to k` is equivalent to `2^(k+1) - 1`.

I couldn't find a decent link describing this sum for these comments, but I did find a very brief description on wikipedia (just the first one in this section): https://en.wikipedia.org/wiki/1_%2B_2_%2B_4_%2B_8_%2B_%E2%8B%AF#Summation

Since the first thing this thread does is the sleep, then at this point in the code `sleep-seconds*2 - 1` = `2^n - 1` if `n` is the number of times the thread has already slept.

Also, I agree that it's not super important that we fit in exactly under the maximum, so the `wait-seconds-max` config could be any value, and you're right we would only pass that max by 1 loop if the config isn't a value equal to `2^n-1`. That could mean this thread could sleep nearly twice as long as the max config (for example if the config were 1024 then thread could end up waiting 1024*2-1=2047 seconds or just over 34 minutes), but that still doesn't seem like that big of a deal for now.